README
NAME
    CPAN::Testers::Common::Utils - Utility functions for CPAN Testers
    modules

VERSION
    This documentation describes version 0.001.

SYNOPSIS
         use CPAN::Testers::Common::Utils ':all';
 
         # NNTP ID <=> GUID mapping
         $guid    = nntp_to_guid( $nntp_id );
         $nntp_id = guid_to_nntp( $guid    );

DESCRIPTION
    This module contains common utility functions for use by other CPAN
    Testers modules

USAGE
  Mapping NNTP IDs to GUIDs
    Legacy CPAN Testers reports were sent via email and made available via
    an NNTP group, C<perl.cpan.testers>. Reports were 'indexed' by their
    NNTP ID. The next generation of CPAN Testers uses a GUID URN to identify
    reports.

    Old reports with an NNTP ID are mapped to GUIDs by replacing the first 8
    hex characters of a common 'base GUID' with a zero-padded decimal
    representation of the NNTP ID.

       XXXXXXXX-b19f-3f77-b713-d32bba55d77f

    Such GUID URNs are visually distinctive and have the nice feature of
    sorting earlier than second-generated report GUIDs based on a timestamp.

    Two translation functions are provided for convenience.

    `nntp_to_guid'
         $guid    = nntp_to_guid( $nntp_id );

    Given a numeric NNTP ID, returns a standard string-form GUID. (No range
    checking is done.) Examples:

       nntp_to_guid( 51432   );  # 00051432-b19f-3f77-b713-d32bba55d77f
       nntp_to_guid( 6171265 );  # 06171265-b19f-3f77-b713-d32bba55d77f

    `guid_to_nntp'
         $guid    = nntp_to_guid( $nntp_id );

    Given a GUID string of the form described above, returns the decimal
    number in the first 8 characaters. (Again, there is no error checking
    that the GUID is properly formatted.) Examples:

       guid_to_nntp( '00051432-b19f-3f77-b713-d32bba55d77f' ); # 51432
       guid_to_nntp( '06171265-b19f-3f77-b713-d32bba55d77f' ); # 6171265

BUGS
    Please report any bugs or feature requests using the CPAN Request
    Tracker web interface at
    http://rt.cpan.org/Dist/Display.html?Queue=CPAN-Testers-Common-Utils

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

SEE ALSO
    *   Data::GUID::Any

AUTHOR
    David A. Golden (DAGOLDEN)

COPYRIGHT AND LICENSE
    Copyright (c) 2010 by David A. Golden. All rights reserved.

    Licensed under Apache License, Version 2.0 (the "License"). You may not
    use this file except in compliance with the License. A copy of the
    License was distributed with this file or you may obtain a copy of the
    License from
    http:E<sol>E<sol>www.apache.orgE<sol>licensesE<sol>LICENSE-2.0

    Files produced as output though the use of this software, shall not be
    considered Derivative Works, but shall be considered the original work
    of the Licensor.

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.



Changes
Revision history for CPAN-Testers-Common-Utils

{{ $NEXT }}

0.003     2015-04-16 23:04:25+02:00 Europe/Berlin

    [Meta]

    - Updated repo and bugtracker information

0.002     2010-06-04 10:37:45 EST5EDT

    - have guid_to_nntp return undef if GUID is not a converted NNTP GUID

    - distribution now managed with Dist::Zilla

0.001 Sun Jan 17 22:34:30 EST 2010

    - Initial release


Build.PL
# Copyright (c) 2010 by David Golden. All rights reserved.
# Licensed under Apache License, Version 2.0 (the "License").
# You may not use this file except in compliance with the License.
# A copy of the License was distributed with this file or you may obtain a
# copy of the License from http://www.apache.org/licenses/LICENSE-2.0

use strict;
use lib 'inc';
eval "require Pod::WikiDoc";
my $class = $@ ? "Module::Build" : "Module::Build::WikiDoc";
eval "require $class";
my $build = $class->new(
    module_name => 'CPAN::Testers::Common::Utils',
    dist_author => 'David Golden <dagolden@cpan.org>',
    license => 'apache',
    create_readme => 1,
    create_makefile_pl  => 'traditional',
    requires => {
        'perl' => '5.006',
    },
    build_requires => {
        'Test::More' => 0.47,
    },
    no_index => { 
      directory => [ qw/ examples inc t xt /],
    },
    meta_add            => {
        resources => {
          repository => "git://github.com/dagolden/cpan-testers-common-utils.git",
          bugtracker => "http://rt.cpan.org/NoAuth/Bugs.html?Dist=CPAN-Testers-Common-Utils",
        },
    },
);
$build->create_build_script;



Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.035.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.17;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Utility functions for CPAN Testers modules",
  "AUTHOR" => "David A Golden <dagolden\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.17"
  },
  "DISTNAME" => "CPAN-Testers-Common-Utils",
  "EXE_FILES" => [],
  "LICENSE" => "apache",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "CPAN::Testers::Common::Utils",
  "PREREQ_PM" => {
    "Exporter" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.004",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => "6.17",
  "File::Spec" => 0,
  "Test::More" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/CPAN/Testers/Common/Utils.pm
use strict;
use warnings;
package CPAN::Testers::Common::Utils;
# ABSTRACT: Utility functions for CPAN Testers modules

our $VERSION = '0.004';

use Exporter ();
our @ISA = qw/Exporter/;
our @EXPORT_OK = qw(
  nntp_to_guid
  guid_to_nntp
);
our %EXPORT_TAGS = (
  all => [@EXPORT_OK]
);

#--------------------------------------------------------------------------#
# NNTP <-> GUID
#--------------------------------------------------------------------------#

# Base GUID generated with:
# Data::UUID->new->create_from_name_str(
#   NameSpace_URL, "http://nntp.x.perl.org/group/perl.cpan.testers"
# );

# Lower case is canonical
my $base_guid = "ed372d00-b19f-3f77-b713-d32bba55d77f";

# strip leading zeros on extraction
my $nntp_re = qr{\A0*([0-9]{1,7})-b19f-3f77-b713-d32bba55d77f$};

sub nntp_to_guid {
  my ($nntp_id) = @_;
  my $guid = $base_guid;
  substr($guid, 0, 8, sprintf("%08d",$nntp_id)); # zero padded
  return $guid;
}

sub guid_to_nntp {
  my ($guid) = @_;
  my ($nntp_id) = $guid =~ $nntp_re;
  return $nntp_id;
}

1;

__END__

=head1 SYNOPSIS

    use CPAN::Testers::Common::Utils ':all';

    # NNTP ID <=> GUID mapping
    $guid    = nntp_to_guid( $nntp_id );
    $nntp_id = guid_to_nntp( $guid    );

=head1 DESCRIPTION

This module contains common utility functions for use by other CPAN Testers
modules

=head1 USAGE

=head2 Mapping NNTP IDs to GUIDs

Legacy CPAN Testers reports were sent via email and made available via an
NNTP group, C<perl.cpan.testers>.  Reports were 'indexed' by their NNTP ID.
The next generation of CPAN Testers uses a GUID URN to identify reports.

Old reports with an NNTP ID are mapped to GUIDs by replacing the first 8
hex characters of a common 'base GUID' with a zero-padded decimal
representation of the NNTP ID.

  XXXXXXXX-b19f-3f77-b713-d32bba55d77f

Such GUID URNs are visually distinctive and have the nice feature of
sorting earlier than second-generated report GUIDs based on a timestamp.

Two translation functions are provided for convenience.

=head3 nntp_to_guid

    $guid    = nntp_to_guid( $nntp_id );

Given a numeric NNTP ID, returns a standard string-form GUID.  (No range
checking is done.) Examples:

  nntp_to_guid( 51432   );  # 00051432-b19f-3f77-b713-d32bba55d77f
  nntp_to_guid( 6171265 );  # 06171265-b19f-3f77-b713-d32bba55d77f

=head3 guid_to_nntp

    $guid    = nntp_to_guid( $nntp_id );

Given a GUID string of the form described above, returns the decimal number
in the first 8 characters.  Examples:

  guid_to_nntp( '00051432-b19f-3f77-b713-d32bba55d77f' ); # 51432
  guid_to_nntp( '06171265-b19f-3f77-b713-d32bba55d77f' ); # 6171265

If the GUID string is not derived from the base GUID, this function 
returns C<undef>.

=head1 SEE ALSO

=for :list
* L<Data::GUID::Any>

=cut

